subsys/bluetooth/mesh/cdb.c

```diff
@@ -0,0 +1,364 @@
+{
+	struct bt_mesh_cdb_subnet *sub;
+
+	if (atomic_test_and_set_bit(bt_mesh_cdb.flags,
+				    BT_MESH_CDB_VALID)) {
+		return -EALREADY;
+	}
+
+	sub = bt_mesh_cdb_subnet_alloc(BT_MESH_KEY_PRIMARY);
+	if (sub == NULL) {
+		return -ENOMEM;
+	}
+
+	memcpy(sub->keys[0].net_key, key, 16);
+	bt_mesh_cdb.iv_index = 0;
+
+	if (IS_ENABLED(CONFIG_BT_SETTINGS)) {
+		bt_mesh_store_cdb();
+		bt_mesh_store_cdb_subnet(sub);
+	}
+
+	return 0;
+}
+
+int bt_mesh_cdb_clear(void)
+{
+	int i;
+
+	atomic_clear_bit(bt_mesh_cdb.flags, BT_MESH_CDB_VALID);
+
+	for (i = 0; i < ARRAY_SIZE(bt_mesh_cdb.nodes); ++i) {
+		if (bt_mesh_cdb.nodes[i].addr != BT_MESH_ADDR_UNASSIGNED) {
+			bt_mesh_cdb_node_del(&bt_mesh_cdb.nodes[i], true);
+		}
+	}
+
+	for (i = 0; i < ARRAY_SIZE(bt_mesh_cdb.subnets); ++i) {
+		if (bt_mesh_cdb.subnets[i].net_idx != BT_MESH_KEY_UNUSED) {
+			bt_mesh_cdb_subnet_del(&bt_mesh_cdb.subnets[i], true);
+		}
+	}
+
+	for (i = 0; i < ARRAY_SIZE(bt_mesh_cdb.app_keys); ++i) {
+		if (bt_mesh_cdb.app_keys[i].net_idx != BT_MESH_KEY_UNUSED) {
+			bt_mesh_cdb_app_key_del(&bt_mesh_cdb.app_keys[i], true);
+		}
+	}
+
+	if (IS_ENABLED(CONFIG_BT_SETTINGS)) {
+		bt_mesh_store_cdb();
+	}
+
+	return 0;
+}
+
+void bt_mesh_cdb_iv_update(u32_t iv_index, bool iv_update)
+{
+	BT_DBG("Updating IV index to %d\n", iv_index);
+
+	bt_mesh_cdb.iv_index = iv_index;
+
+	atomic_set_bit_to(bt_mesh_cdb.flags, BT_MESH_CDB_IVU_IN_PROGRESS,
+			  iv_update);
+
+	if (IS_ENABLED(CONFIG_BT_SETTINGS)) {
+		bt_mesh_store_cdb();
+	}
+}
+
+struct bt_mesh_cdb_subnet *bt_mesh_cdb_subnet_alloc(u16_t net_idx)
+{
+	struct bt_mesh_cdb_subnet *sub;
+	int i;
+
+	if (bt_mesh_cdb_subnet_get(net_idx) != NULL) {
+		return NULL;
+	}
+
+	for (i = 0; i < ARRAY_SIZE(bt_mesh_cdb.subnets); ++i) {
+		sub = &bt_mesh_cdb.subnets[i];
+
+		if (sub->net_idx != BT_MESH_KEY_UNUSED) {
+			continue;
+		}
+
+		sub->net_idx = net_idx;
+
+		return sub;
+	}
+
+	return NULL;
+}
+
+void bt_mesh_cdb_subnet_del(struct bt_mesh_cdb_subnet *sub, bool store)
+{
+	BT_DBG("NetIdx 0x%03x store %u", sub->net_idx, store);
+
+	if (IS_ENABLED(CONFIG_BT_SETTINGS) && store) {
+		bt_mesh_clear_cdb_subnet(sub);
+	}
+
+	sub->net_idx = BT_MESH_KEY_UNUSED;
+	memset(sub->keys, 0, sizeof(sub->keys));
+}
+
+struct bt_mesh_cdb_subnet *bt_mesh_cdb_subnet_get(u16_t net_idx)
+{
+	int i;
+
+	for (i = 0; i < ARRAY_SIZE(bt_mesh_cdb.subnets); ++i) {
+		if (bt_mesh_cdb.subnets[i].net_idx == net_idx) {
+			return &bt_mesh_cdb.subnets[i];
+		}
+	}
+
+	return NULL;
+}
+
+u8_t bt_mesh_cdb_subnet_flags(const struct bt_mesh_cdb_subnet *sub)
+{
+	u8_t flags = 0x00;
+
+	if (sub && sub->kr_flag) {
+		flags |= BT_MESH_NET_FLAG_KR;
+	}
+
+	if (atomic_test_bit(bt_mesh_cdb.flags, BT_MESH_CDB_IVU_IN_PROGRESS)) {
+		flags |= BT_MESH_NET_FLAG_IVU;
+	}
+
+	return flags;
+}
+
+struct bt_mesh_cdb_node *bt_mesh_cdb_node_alloc(const u8_t uuid[16], u16_t addr,
+						u8_t num_elem, u16_t net_idx)
+{
+	int i;
+
+	BT_DBG("");
+
+	if (addr == BT_MESH_ADDR_UNASSIGNED) {
```

Comment:
To reallocate addresses to nodes that are rejoining network we (you) could add something like:

```
	for (i = 0; i < ARRAY_SIZE(bt_mesh_cdb.nodes); i++) {
		struct bt_mesh_cdb_node *node = &bt_mesh_cdb.nodes[i];

		if (memcmp(node->uuid, uuid, 16)) {
			return node;
		}
	}
```
This should stop the same piece of hardware filling up the CDB with multiple entries. Its been great for testing but in the wild this will quickly become a problem.

Billy..

Reply:
Just had a thought, if the node is unprovisioned and is reassigned the same address then the CDB entry needs to be set to  'not configured' as well. Oops should have noticed first time.

Reply:
This is the way the nRF Mesh Android app works :+1: 

We might want to change the mesh sample to generate a unique UUID for each device though, currently it's hardcoded to one value for all deployments, which would break this.

Reply:
or an easier option is to make it a configuration option: MESH_CDB_NODE_REALLOCATE

Reply:
I think the already is something that assigns the uuid based on chip id in the mesh sample or the are BTT UUID creation functions in uuid.h

Reply:
Here is something that is in the nrf board samples, could easily be added to the samples if needed.
```
	/* Use identity address as device UUID */
	if (bt_le_oob_get_local(BT_ID_DEFAULT, &oob)) {
		printk("Identity Address unavailable\n");
	} else {
		memcpy(dev_uuid, oob.addr.a.val, 6);
	}
```
It may cause issues with none Nordic boards but I just don't have any to test.

Reply:
@WilliamGFish Not sure what you want to do, but to get the identity address it would be better to use `bt_id_get`